File: algorithms/000988-smallest-string-starting-from-leaf/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        def DFS(root,str):
            if root is not None:
                if root.left is None and root.right is None:
                    str+=chr(root.val+ord('a'))
                    RevStr = str[::-1]
                    #print("RevStr {}".format(RevStr))
                    return RevStr
                else:
                    str1=DFS(root.left,str+chr(root.val+ord('a')))
                    str2=DFS(root.right,str+chr(root.val+ord('a')))
                    #print("str1 {} , str2 {}".format(str1,str2))
                    if(str1 and str2):
                        return min(str1,str2)
                    elif (str1):
                        return str1
                    else:
                        return str2
            else:
                return ""
        return DFS(root,"")

File: algorithms/000988-smallest-string-starting-from-leaf/solution.php
<?php
/**
 * Definition for a binary tree node.
 */

class TreeNode
{
    public mixed $val = null;
    public mixed $left = null;
    public mixed $right = null;

    function __construct($val = 0, $left = null, $right = null)
    {
        $this->val = $val;
        $this->left = $left;
        $this->right = $right;
    }
}

class Solution
{

    /**
     * @param TreeNode $root
     * @return String
     */
    function smallestFromLeaf(TreeNode $root): string
    {
        $ans = $this->dfs($root, new SplStack(), array());
        sort($ans);
        return $ans[0];
    }

    public function dfs($root, $stack, $ans)
    {
        if ($root == null) return $ans;
        $stack->push($root);
        if ($root->left == null && $root->right == null) {
            $sb = "";
            foreach ($stack as $n) {
                $sb .= chr(97 + $n->val);
            }
            $ans[] = strrev($sb);
        }

        $this->dfs($root->left, $stack, $ans);
        while ($stack->top()->right != $root->right) $stack->pop();
        $this->dfs($root->right, $stack, $ans);
        return $ans;
    }
}